dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilder.java

```diff
@@ -124,70 +156,90 @@ public static DependencyGraph getCompleteDependencies(Artifact artifact)
+    Queue<LevelOrderQueueItem> queue = new ArrayDeque<>();
+    queue.add(new LevelOrderQueueItem(node, new Stack<>()));
+    while (!queue.isEmpty()) {
+      LevelOrderQueueItem item = queue.poll();
+      DependencyNode dependencyNode = item.dependencyNode;
+      DependencyPath forPath = new DependencyPath();
+      Stack<DependencyNode> parentNodes = item.parentNodes;
+      parentNodes.forEach(parentNode -> forPath.add(parentNode.getArtifact()));
+      if (dependencyNode.getArtifact() != null) {
+        // When requesting dependencies of 2 or more artifacts, root is null
+        forPath.add(dependencyNode.getArtifact());
+        graph.addPath(forPath);
+        parentNodes.push(dependencyNode);
+      }
+      for (DependencyNode childNode : dependencyNode.getChildren()) {
+        queue.add(new LevelOrderQueueItem(childNode, (Stack<DependencyNode>) parentNodes.clone()));
       }
     }
-    graph.addPath(forPath);
-    
-    for (DependencyNode child : current.getChildren()) {
-      if (!"system".equals(child.getDependency().getScope())) {
-        try {
-          child = resolveCompileTimeDependencies(child.getArtifact());
-          // somehow we've got an infinite recursion here
-          // requires equals
-          if (path.contains(child)) {
-            System.err.println("Infinite recursion resolving " + current);
-            System.err.println("Likely cycle in " + forPath);              
-            System.err.println("Child " + child);
-          } else {
-            fullPreorder((Stack<DependencyNode>) path.clone(), child, graph);
+  }
+
+  private static void fullLevelOrder(DependencyNode node, DependencyGraph graph)
+      throws DependencyCollectionException, DependencyResolutionException {
+    Queue<LevelOrderQueueItem> queue = new ArrayDeque<>();
+    queue.add(new LevelOrderQueueItem(node, new Stack<>()));
+    while (!queue.isEmpty()) {
+      LevelOrderQueueItem item = queue.poll();
+      DependencyNode dependencyNode = item.dependencyNode;
+      DependencyPath forPath = new DependencyPath();
+      Stack<DependencyNode> parentNodes = item.parentNodes;
+      parentNodes.forEach(parentNode -> forPath.add(parentNode.getArtifact()));
+      if (dependencyNode.getArtifact() != null) {
+        // When requesting dependencies of 2 or more artifacts, root is null
+        forPath.add(dependencyNode.getArtifact());
+        if (parentNodes.contains(dependencyNode)) {
+          System.err.println("Infinite recursion resolving " + dependencyNode);
+          System.err.println("Likely cycle in " + parentNodes);
+          System.err.println("Child " + dependencyNode);
+          return;
+        }
+        parentNodes.push(dependencyNode);
+        graph.addPath(forPath);
+
+        if (!"system".equals(dependencyNode.getDependency().getScope())) {
+          try {
+            dependencyNode = resolveCompileTimeRootDependencies(dependencyNode.getArtifact());
```

Comment:
I can't tell what the difference of intent is between `levelOrder` and `fullLevelOrder` but this seems to be an important difference - what does this line accomplish?

Reply:
Added javadoc to `levelOrder`.

`resolveCompileTimeRootDependencies` itself doesn't dig further when a dependency is `optional: true`. The function `fullLevelOrder`, on the other hand, calls `resolveCompileTimeRootDependencies` every time it finds a dependency so that it can fetch dependencies of a `optional: true` dependency. 

Example is noted in

> jaxen has dom4j as dependency (optional: true

https://github.com/GoogleCloudPlatform/cloud-opensource-java/issues/162#issuecomment-430063293